[inductor] Fix nested indirect indexing case for index_propagation

ColinPeppler opened the pull request:
Tries to fix #127677.

# Context

Just as @peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
When creating `indirect` symbols from fallback, specify its range to be `[-size, size -1]` to avoid a lookup error with `indirectX`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128378



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang

ColinPeppler committed:
[inductor] Fix nested indirect indexing case for index_propagation

[ghstack-poisoned]

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler commented:
So, if we still include the bounds check instead of returning `expr` immediately, we'll have two bounds checks.

```
tmp4 = tl.where(tmp3, tmp2, tmp0)
# One from codegen pass
tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")
# One from IndexPropagation -- this one seems like the stricter bounds check (but safer?). 
tl.device_assert(((0 <= tmp4) & (tmp4 < 32)) | ~(xmask), "index out of bounds: 0 <= tmp4 < 32")
```

Thinking about whether it makes sense to avoid the bounds check from the codegen pass and use the one from IndexPropagation.

lezcano commented on torch/_inductor/index_propagation.py:
```diff
@@ -322,6 +323,13 @@ def indirect_indexing(
 
+            if free_symbol_is_type(expr, SymT.INDIRECT):
+                # This is the nested indirect indexing case:
+                # a = ops.indirect_indexing(...)
+                # b = ops.index_expr(a, ...)
+                # c = ops.indirect_indexing(b, ...)
```
In this case, it doesn't seem like we can meaningfully propagate a sympy expression, so don't we want to fallback?

peterbell10 replied:
No, we should be able to propagate the sympy expression without issue. We should just fix the `_maybe_evaluate_static` call so it can handle `indirect0`.

Ideally we would keep track of the bounds for `ops.indirect_indexing` since these do have known bounds. But at the very least we should treat it as unbounded.

peterbell10 replied:
For example
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, torch.int32)
c = ops.add(b, ops.constant(1, torch.int32))
d = ops.indirect_indexing(c, ...)
```

`d = indirect1` even though we know that `indirect1 = indirect0 + 1`

We can simplify to:
```
a = ops.indirect_indexing(...)
d = a + 1
```

where `d = indirect0 + 1`.

lezcano replied:
To do this we should:
- Make sure that the shapes of the tow indirect_indexings are the same
- Make sure we are not wrapping the variable twice (I think we may be currently wrapping it twice).

It's not clear to me whether we can do the first thing in the IndexPropagation pass tho.

peterbell10 replied:
We should propagate the bounds from the first `indirect0` into the reasoning, so we might be able to elide the wrapping.

In general though, we might need to have two wrappings and two different sizes because each `ops.indirect_indexing` call might have come from a different tensor which may have different shapes and each are wrapped independently.

lezcano replied:
indirect0 variables barely ever have bounds, as loads do not have bounds, and indirect indexing is often performed on information that comes from a load.

But yeah, I guess that in general it makes sense for the first indirect to be wrapped and so on at a codegen level, and then we can have the second indirect wrapped here.

peterbell10 replied:
> indirect0 variables barely ever have bounds,

On the contrary, `indirect0` always has a bound. `ops.indirect_indexing(var, size)` has the bound [0, size) as checked by the runtime assert.

lezcano replied:
ugh, right.

lezcano commented on torch/_inductor/index_propagation.py:
```diff
@@ -322,6 +323,13 @@ def indirect_indexing(
+                # a = ops.indirect_indexing(...)
+                # b = ops.index_expr(a, ...)
+                # c = ops.indirect_indexing(b, ...)
+                return Where(expr < 0, expr + size, expr)
```
use `wrap_expr` to cover the (rare) case when we can do better than wrapping.

lezcano reviewed (approved):
Just use `wrap_expr`, but everything else SGTM

lezcano reviewed (changes requested):
Actually, we still need to emit the check_bounds here unless we can prove that the sizes are the same, no? And in that case, we probably also want to emit it and just CSE it at codegen level.

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"

Tries to fix #127677.

# Context

Just as peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
We can probably skip the check to prove `indirect0`'s bounds because its purpose is to add a bounds check as a device side assert. Thankfully, we already do this in the codegen pass.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/codegen/common.py#L1730-L1733





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler commented:
We can't prove the sizes are the same i.e. s0 == 32.

So, now we'll emit another check_bounds() which gives us two of them:
```
# First check_bounds() from torch.arange(repeats.numel())
tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")
# Second check_bounds() from torch.index_select(x ...)
tl.device_assert((tmp4 < 32) | ~(xmask), "index out of bounds: tmp4 < 32")
```

IMO the double check_bounds on `tmp4` does feel a bit weird. If anything, the second check is all that's necessary, but I'm unsure how to prevent the first check.

lezcano commented on torch/_inductor/index_propagation.py:
```diff
@@ -322,6 +324,18 @@ def indirect_indexing(
+            # b = ops.index_expr(a, ...)
+            # c = ops.indirect_indexing(b, ...)
+            indirect_var_to_default_range = tuple(
+                (sym, self.shape_env._default_unspecified_value_range())
```
So, in the pattern above, you are going to hit this function twice. The first one you will not be able to remove the `indirect_indexing`, but you will be able to set proper ranges for it. Then, in the second one, the ranges will already be set, and you will be able to just use those ranges.

The invariant here is that, if you fallback, you will be able to get the result of the fallback, (which is a sympy.Symbol) and add its range as `[-size, size-1]` (as it's not wrapped at this point). Then any call after it should work as expected.
You can even assert that the returned symbol is not in `var_to_range` before adding it.

ColinPeppler commented:
I realize that `indirect0` was intially created with the assumption it's non-negative. This means `0 <= expr` will always evaluate to true, although it can be negative (atleast until `wrap_expr()`).

https://github.com/pytorch/pytorch/blob/3bc2004f9123a32f381ef64202252d59109507f3/torch/_inductor/utils.py#L596-L605

So, I tossed that assumption out since `indirect0` can be negative until it gets wrapped.

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"

Tries to fix #127677.

# Context

Just as peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
Add `indirectX  symbols with a default range (-inf, +inf) to `self.var_to_range` to avoid a lookup error with `indirectX`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"

Tries to fix #127677.

# Context

Just as peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
Add `indirectX  symbols with a default range (-inf, +inf) to `self.var_to_range` to avoid a lookup error with `indirectX`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented on torch/_inductor/index_propagation.py:
```diff
@@ -345,4 +346,12 @@ def wrap_expr(expr):
+        indirect_var = self.fallback(
+            "indirect_indexing", (index, size, check), {}
+        ).value
+        if indirect_var not in self.var_to_range:
+            lower, upper = -upper_bound(size), upper_bound(size) - 1
+            indirect_range = (indirect_var, ValueRanges(lower, upper))
+            self.var_to_range = self.var_to_range + (indirect_range,)
```
Can this be turned into
```suggestion
        assert indirect_var not in self.var_to_range
        lower, upper = -upper_bound(size), upper_bound(size) - 1
        indirect_range = (indirect_var, ValueRanges(lower, upper))
        self.var_to_range = self.var_to_range + (indirect_range,)
```
?

ColinPeppler replied:
sure!

lezcano reviewed (commented):
The approach largely looks good. Just one question.

Also, it's a pity that, in general, we have to drop the `is_positive` assumption from SymPy, but it is true that it was incorrect as written.

lezcano commented on torch/_inductor/ir.py:
```diff
@@ -8012,7 +8012,10 @@ def add_submodule(self, block, prefix):
+        from torch.utils._sympy.symbol import make_symbol
+
+        # Note: indirect index variables can be negative, positive or 0.
+        var = make_symbol(SymT.INDIRECT, len(self.indirect_vars), integer=True)
```
Ah, right, we do the wrapping within `indirect_indexing` in codegen, so I guess the positive condition is correct here.

ColinPeppler replied:
That makes sense, I'll change that back because we know it must be non-negative after wrapping it.

> wrapping of negative values should happen before the real value is assigned to the symbolic value
Btw I'm trying to understand what this looks like. Can you help me understand? Thanks!

lezcano replied:
see the codegen for indirect_index in common.py

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"

Tries to fix #127677.

# Context

Just as peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
When creating `indirect` symbols from fallback, specify its range to be `[-size, size -1]` to avoid a lookup error with `indirectX`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/index_propagation.py:
```diff
@@ -345,4 +346,15 @@ def wrap_expr(expr):
+            indirect_var not in self.var_to_range
+        ), f"{indirect_var} should've been created in the fallback."
+
+        lower, upper = -upper_bound(size), upper_bound(size) - 1
```
```suggestion
        lower, upper = sympy.Integer(0), upper_bound(size) - 1
```

lezcano replied:
and even inline it, if you want

ColinPeppler committed:
Update on "[inductor] Fix nested indirect indexing case for index_propagation"

Tries to fix #127677.

# Context

Just as peterbell10 pointed out, we have the following scenario:
```
a = ops.indirect_indexing(...)
b = ops.index_expr(a, ...)
c = ops.indirect_indexing(b, ...)
```

We can repro this as:
```
def forward(self, arg0_1, arg1_1, arg2_1):
    iota = torch.ops.prims.iota.default(arg0_1, start = 0, step = 1, index=0),
    repeat_interleave = torch.ops.aten.repeat_interleave.Tensor(arg1_1);
    index = torch.ops.aten.index.Tensor(iota, [repeat_interleave]);
    index_1 = torch.ops.aten.index.Tensor(arg2_1, [index]);
    return (index_1,)
```

which should generate a JIT py file like this:
```
def triton_poi_fused_index_select_0(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, XBLOCK : tl.constexpr):
    ...
    tmp0 = tl.load(in_ptr0 + (x1), xmask, eviction_policy='evict_last')
    tmp1 = ks0
    tmp2 = tmp0 + tmp1
    tmp3 = tmp0 < 0
    tmp4 = tl.where(tmp3, tmp2, tmp0)
    # check_bounds()
    tl.device_assert(((0 <= tmp4) & (tmp4 < ks0)) | ~(xmask), "index out of bounds: 0 <= tmp4 < ks0")

def call():
  arg0_1, arg1_1, arg2_1 = args
  buf1 = aten.repeat_interleave.Tensor(arg1_1)
  buf4 = empty_strided_cuda((u0, 64), (64, 1))
  triton_poi_fused_index_select_0.run(
    buf1, arg2_1, buf4, s0, 
    triton_poi_fused_index_select_0_xnumel, 
    grid=grid(triton_poi_fused_index_select_0_xnumel), 
    stream=stream0)
```

# Issue
In our `IndexPropagation.indirect_indexing()` call we have `expr=indirect0` which is spawned in `LoopBodyBlock.indirect_indexing()`.
https://github.com/pytorch/pytorch/blob/3b555ba47713d489975a9bb6cb6c31975f805e3f/torch/_inductor/ir.py#L8154-L8160

When we try to see if we can prove its bounds, we fail because `indirect0` isn't in `var_ranges`.

# Approach
When creating `indirect` symbols from fallback, specify its range to be `[-size, size -1]` to avoid a lookup error with `indirectX`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang

[ghstack-poisoned]

lezcano reviewed (approved):
Great, thank you!

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot merge -i
